jvm-packages/xgboost4j/src/main/scala/ml/dmlc/xgboost4j/scala/rabit/util/LinkMap.scala

```diff
@@ -0,0 +1,136 @@
+/*
+ Copyright (c) 2014 by Contributors
+
+ Licensed under the Apache License, Version 2.0 (the "License");
+ you may not use this file except in compliance with the License.
+ You may obtain a copy of the License at
+
+ http://www.apache.org/licenses/LICENSE-2.0
+
+ Unless required by applicable law or agreed to in writing, software
+ distributed under the License is distributed on an "AS IS" BASIS,
+ WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ See the License for the specific language governing permissions and
+ limitations under the License.
+ */
+
+package ml.dmlc.xgboost4j.scala.rabit.util
+
+import java.nio.{ByteBuffer, ByteOrder}
+
+/**
+  * The assigned rank to a connecting Rabit worker, along with the information of the ranks of
+  * its linked peer workers, which are critical to perform Allreduce.
+  * When RabitWorkerHandler delegates "start" or "recover" commands from the connecting worker
+  * client, RabitTrackerHandler utilizes LinkMap to figure out linkage relationships, and respond
+  * with this class as a message, which is later encoded to byte string, and sent over socket
+  * connection to the worker client.
+  *
+  * @param rank assigned rank (ranked by worker connection order: first worker connecting to the
+  *             tracker is assigned rank 0, second with rank 1, etc.)
+  * @param neighbors ranks of neighboring workers in a tree map.
+  * @param ring ranks of neighboring workers in a ring map.
+  * @param parent rank of the parent worker.
+  */
+case class AssignedRank private[scala](rank: Int, neighbors: Seq[Int],
+                                       ring: (Int, Int), parent: Int) {
+  /**
+    * Encode the AssignedRank message into byte sequence for socket communication with Rabit worker
+    * client.
+    * @param worldSize the number of total distributed workers. Must match `numWorkers` used in
+    *                  LinkMap.
+    * @return a ByteBuffer containing encoded data.
+    */
+  def toByteBuffer(worldSize: Int): ByteBuffer = {
+    val buffer = ByteBuffer.allocate(4 * (neighbors.length + 6)).order(ByteOrder.nativeOrder())
+    buffer.putInt(rank).putInt(parent).putInt(worldSize).putInt(neighbors.length)
+    // neighbors in tree structure
+    neighbors.foreach { n => buffer.putInt(n) }
+    buffer.putInt(if (ring._1 != -1 && ring._1 != rank) ring._1 else -1)
+    buffer.putInt(if (ring._2 != -1 && ring._2 != rank) ring._2 else -1)
+
+    buffer.flip()
+    buffer
+  }
+}
+
+class LinkMap private[scala](numWorkers: Int) {
```

Comment:
you need to put private[scala] in front of class, otherwise, you're just make cosntructor private instead of the whole class